src/components/shapes/index.js

```diff
@@ -288,23 +294,72 @@ function updateShape(gd, index, opt, value) {
     gd._fullLayout.shapes[index] = options;
 
     var attrs = {
-            'data-index': String(index),
+            'data-index': index,
             'fill-rule': 'evenodd',
             d: shapePath(gd, options)
         },
         clipAxes = (options.xref + options.yref).replace(/paper/g, '');
 
     var lineColor = options.line.width ? options.line.color : 'rgba(0,0,0,0)';
 
-    var path = gd._fullLayout._shapelayer.append('path')
-        .attr(attrs)
-        .style('opacity', options.opacity)
-        .call(Color.stroke, lineColor)
-        .call(Color.fill, options.fillcolor)
-        .call(Drawing.dashLine, options.line.dash, options.line.width);
+    if(options.layer !== 'below') {
+        drawShape(gd._fullLayout._shapeUpperLayer);
+    }
+    else if(options.xref === 'paper' && options.yref === 'paper') {
+        drawShape(gd._fullLayout._shapeLowerLayer);
+    } else {
+        forEachSubplot(gd, function(plotinfo) {
+            if(isShapeInSubplot(gd, options, plotinfo.id)) {
+                drawShape(plotinfo.shapelayer);
+            }
+        });
+    }
+
+    return;
```

Comment:
I understand. I'm not saying that it's a bad habit. It's really a matter of consistency with the rest of the source code.